### PR TITLE
Ruby: Add Insecure Randomness Query 

### PR DESCRIPTION
**Pull request by maikypedia**
This pull request adds a query for Insecure Randomness.

Looking forward to your suggestions.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/**
+ * Provides default sources, sinks, and sanitizers for reasoning about bypass of
+ * sensitive action guards, as well as extension points for adding your own.
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll:3]: `about random values that are not cryptographically secure`?

File: ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+private import codeql.ruby.CFG
+private import codeql.ruby.AST
+private import codeql.ruby.DataFlow
+private import codeql.ruby.security.SensitiveActions
+private import codeql.ruby.Concepts
+private import codeql.ruby.ApiGraphs
+
+/**
+ * Provides default sources, sinks, and sanitizers for reasoning about bypass of
+ * sensitive action guards, as well as extension points for adding your own.
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll:15]: ditto

File: ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A simple random number generator that is not cryptographically secure.
+   */
+  class DefaultSource extends Source, DataFlow::CallNode {
+    DefaultSource() {
+      this = API::getTopLevelMember("Random").getAMethodCall(["rand"])
+      or
+      this.asExpr().getExpr() instanceof UnknownMethodCall and
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll:40]: If you `import codeql.ruby.frameworks.core.Kernel` then you can just do
```ql
or this.(Kernel::KernelMethodCall).getMethodName() = "rand"
```

File: ruby/ql/lib/codeql/ruby/security/InsecureRandomnessQuery.qll [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+private import codeql.ruby.DataFlow
+private import codeql.ruby.TaintTracking
+import InsecureRandomnessCustomizations::InsecureRandomness
+
+/**
+ * A taint-tracking configuration for detecting Insecure Randomness vulnerabilities.
+ * DEPRECATED: Use `InsecureRandomnessFlow`
+ */
+deprecated class Configuration extends TaintTracking::Configuration {
```
Comment (hmac) [inline ruby/ql/lib/codeql/ruby/security/InsecureRandomnessQuery.qll:14]: I don't think there's any need to add a deprecated class. We should remove this.

**Review by hmac — CHANGES_REQUESTED** [review]
This looks good! I have some minor comments, and I've kicked off a variant analysis run to see what results we get.

**Comment by hmac** [comment]
There are also two code scanning warnings that should be fixed up.

**Comment by hmac** [comment]
It looks like `ruby/ql/lib/codeql/ruby/security/InsecureRandomnessCustomizations.qll` is not formatted correctly - run `codeql query format` on it to resolve that.

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+<recommendation>
+<p>
+When generating values for use in security-sensitive contexts, it's essential to utilize a 
+cryptographically secure pseudo-random number generator. As a general guideline, a value 
+should be deemed "security-sensitive" if its predictability would empower an attacker to 
+perform actions that would otherwise be beyond their reach. For instance, if an attacker could 
+predict a newly generated user's random password, they would gain unauthorized access to that user's 
+account.
+
+For Ruby, `SecureRandom` provides a cryptographically secure pseudo-random number generator.
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:27]:
```suggestion
For Ruby, <code>SecureRandom</code> provides a cryptographically secure pseudo-random number generator. 
```

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+<p>
+When generating values for use in security-sensitive contexts, it's essential to utilize a 
+cryptographically secure pseudo-random number generator. As a general guideline, a value 
+should be deemed "security-sensitive" if its predictability would empower an attacker to 
+perform actions that would otherwise be beyond their reach. For instance, if an attacker could 
+predict a newly generated user's random password, they would gain unauthorized access to that user's 
+account.
+
+For Ruby, `SecureRandom` provides a cryptographically secure pseudo-random number generator. 
+`rand` is not cryptographically secure, and should be avoided in security contexts.
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:28]:
```suggestion
<code>rand</code> is not cryptographically secure, and should be avoided in security contexts. 
```

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+When generating values for use in security-sensitive contexts, it's essential to utilize a 
+cryptographically secure pseudo-random number generator. As a general guideline, a value 
+should be deemed "security-sensitive" if its predictability would empower an attacker to 
+perform actions that would otherwise be beyond their reach. For instance, if an attacker could 
+predict a newly generated user's random password, they would gain unauthorized access to that user's 
+account.
+
+For Ruby, `SecureRandom` provides a cryptographically secure pseudo-random number generator. 
+`rand` is not cryptographically secure, and should be avoided in security contexts. 
+For contexts which are not security sensitive, Random may be preferable as it has a more convenient
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:29]:
```suggestion
For contexts which are not security sensitive, <code>Random</code> may be preferable as it has a more convenient 
```

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+</p>
+</recommendation>
+
+<example>
+<p>
+The following examples show different ways of generating a password.
+</p>
+
+<p>The first example uses `Random.rand()` which is not for security purposes</p>
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:40]:
```suggestion
<p>The first example uses <code>Random.rand()</code> which is not for security purposes</p>
```

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+<sample src="examples/InsecureRandomnessBad.rb" />
+
+<p>In the second example, the password is generated using `SecureRandom.random_bytes` which is a
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:44]:
```suggestion
<p>In the second example, the password is generated using <code>SecureRandom.random_bytes</code> which is a 
```
```suggestion
<p>In the second example, the password is generated using <code>SecureRandom.random_bytes()</code> which is a 
```

File: ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+<p>In the second example, the password is generated using `SecureRandom.random_bytes` which is a 
+cryptographically secure method.</p>
+
+<sample src="examples/InsecureRandomnessGood.rb" />
+</example>
+
+<references>
+<li>Wikipedia: <a href="http://en.wikipedia.org/wiki/Pseudorandom_number_generator">Pseudo-random number generator</a>.</li>
+<li>Common Weakness Enumeration: <a href="https://cwe.mitre.org/data/definitions/338.html">CWE-338</a>.</li>
```
Comment (intrigus-lgtm) [inline ruby/ql/src/experimental/insecure-randomness/InsecureRandomness.qhelp:52]: This should be added automatically based on the tags in the `.ql` file I think.

**Review by hmac — APPROVED** [review]

**Pull request merged**